File: 백준/Silver/5567. 결혼식/결혼식.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];

rl.on('line', (line) => {
    // 입력 받는 공간
    input.push(line.trim());
}).on('close', () => {
    // 문제를 풀고 출력하는 공간
    let n = Number(input.shift());
    let m = Number(input.shift());
    let line = input.map((v) => v.split(' ').map(Number));

    let graph = Array.from({length : n + 1}, () => []);
    for(let [a, b] of line) {
        graph[a].push(b);
        graph[b].push(a);
    }

    for(let i = 0; i <= n; i++) {
        graph[i].sort((a,b) => a-b);
    }

    let visited = Array(n + 1).fill(false);
    const bfs = (v, visited) => {
        let depth = 0;
        let cnt = 0;

        let q = [];
        q.push([v, depth]);

        while(q.length) {
            let [start, depth] = q.shift();
            visited[start] = true;

            if(depth == 2)
                break;
            
            for(let i = 0; i < graph[start][i]; i++) {
                let next = graph[start][i];
                if(!visited[next]) {
                    cnt++;
                    q.push([next, depth+1]);
                    visited[next] = true;
                }
            }
        }

        return cnt;
    };

    let cnt = 0;
    cnt = Math.max(cnt, bfs(1, visited));
    console.log(cnt);
});

File: 백준/Silver/5567. 결혼식/결혼식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

    static ArrayList<Integer>[] graph;
    static boolean[] visited;
    static boolean[] check;
    static int result;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        graph = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            String[] input = br.readLine().split(" ");
            int a = Integer.parseInt(input[0]);
            int b = Integer.parseInt(input[1]);

            graph[a].add(b);
            graph[b].add(a);
        }

        for (int i = 0; i <= n; i++) {
            Collections.sort(graph[i]);
        }

        result = 0;
        visited = new boolean[n + 1];
        check = new boolean[n + 1];
        dfs(1, 0);
        if (result == 0)
            System.out.println(result);
        else
            System.out.println(result - 1);
    }

    static void dfs(int start, int depth) {
        if (depth == 2)
            return;

        if (!visited[start]) {
            visited[start] = true;

            for (int i = 0; i < graph[start].size(); i++) {
                int next = graph[start].get(i);
                // visited[next] = true;
                dfs(next, depth + 1);

                if (!check[next]) {
                    result++;
                    check[next] = true;
                }
            }
        }
    }
}
